### PR TITLE
Paired end python

### PR DESCRIPTION
**Pull request by ebolyen**
Based on #34 

I noticed that `maxEE` seems to be a float, so I changed the registration to indicate that. I'm not confident on the defaults (copied them from the R script) so if special attention could be paid to those, it would be greatly appreciated!

### CODE REVIEW

File: q2_dada2/_denoise.py [hunk 1]
```diff
@@ -6,28 +6,93 @@
...
+            pass
+    if line_count < 2:
+        raise ValueError("No features remain after denoising. "
+                         "Try adjusting your arguments.")
+
+
+def denoise_single(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
+                   trunc_len: int, trim_left: int, max_ee: float=2.5,
+                   trunc_q: int=2, n_threads: int=1,
+                   n_reads_learn: int=1000000, hashed_feature_ids: bool=True
```
Comment (ebolyen) [inline q2_dada2/_denoise.py]: Not sure about the defaults.

File: q2_dada2/_denoise.py [hunk 2]
```diff
@@ -6,28 +6,93 @@
...
+                (skbio.DNA(id_, metadata={'id': id_})
+                 for id_ in table.ids(axis='observation')))
+    return table, rep_sequences
+
+
+def denoise_paired(demultiplexed_seqs: SingleLanePerSamplePairedEndFastqDirFmt,
+                   trunc_len_f: int, trunc_len_r: int,
+                   trim_left_f: int, trim_left_r: int,
+                   max_ee: float=2.5, trunc_q: int=2, n_threads: int=1,
+                   n_reads_learn: int=1000000, hashed_feature_ids: bool=True
```
Comment (ebolyen) [inline q2_dada2/_denoise.py]: same here, what should the defaults be?

**Comment by ebolyen** [comment]
Looks like I'm missing a recent commit, I'll rebase

File: q2_dada2/_denoise.py (hunk 1 above)
Comment (benjjneb) [inline q2_dada2/_denoise.py]: I would probably default `max_ee` to `2` and `trim_left` to `0` (under the assumption discussed in #32 that primers are already trimmed).

Also, default for `n_threads` should be chosen based on what you all want as a default - single-threaded (`1`) or detect and use all available cores (`0`, and should explain meaning of that in user-facing docs).

File: q2_dada2/_denoise.py (hunk 2 above)
Comment (benjjneb) [inline q2_dada2/_denoise.py]: As above.

File: q2_dada2/_denoise.py (hunk 1 above)
Comment (ebolyen) [inline q2_dada2/_denoise.py]: Thanks @benjjneb! I think single-threaded is probably the way to go for the moment, and we should be able to tie this parameter into more interesting things in the future.

Is `n_reads_learn` a good value? Should it be `0`?

Comment (benjjneb) [inline q2_dada2/_denoise.py]: 1M is a good default for `n_reads_learn`.

Def not 0, that will cause the script to learn the error rates from just the first sample, which is way too few reads in the most common Miseq protocols.

Comment (ebolyen) [inline q2_dada2/_denoise.py]: So if I understand this correctly, 0 is all reads from only the first sample, and any other number is randomly sampled from all reads of all samples?

Comment (benjjneb) [inline q2_dada2/_denoise.py]: The script will use the first X samples, where X is at least 1, and the number of reads in X is at least greater than `n_reads_learn`, to learn the error rates.

So `n_reads_learn=0` will result in just 1 sample, the first sample, being used to estimate the error rates.

*`n_reads_learn` is really a minimum number of reads, not an exact number*

Comment (ebolyen) [inline q2_dada2/_denoise.py]: Interesting, thanks for the clarification!

File: q2_dada2/_denoise.py [hunk 3]
```diff
@@ -6,28 +6,93 @@
...
-            trunc_len: int, trim_left: int, max_ee: int=2, truncq: int=2,
-            hashed_feature_ids: bool=True) -> (biom.Table, DNAIterator):
+def _check_featureless_table(fp):
+    with open(fp) as fh:
+        # There is a comment line and a header before the feature data
+        for line_count, _ in zip(range(3), fh):
+            pass
+    if line_count < 2:
+        raise ValueError("No features remain after denoising. "
+                         "Try adjusting your arguments.")
```
Comment (gregcaporaso) [inline q2_dada2/_denoise.py]: *arguments* to *truncation and trim parameter settings*

File: q2_dada2/_denoise.py [hunk 4]
```diff
@@ -6,28 +6,93 @@
...
+                   trunc_len: int, trim_left: int=0, max_ee: float=2.0,
+                   trunc_q: int=2, n_threads: int=1,
+                   n_reads_learn: int=1000000, hashed_feature_ids: bool=True
+                   ) -> (biom.Table, DNAIterator):
     with tempfile.TemporaryDirectory() as temp_dir_name:
         biom_fp = os.path.join(temp_dir_name, 'output.tsv.biom')
-        cmd = ['run_dada.R', str(demultiplexed_seqs), biom_fp,
-               str(trunc_len), str(trim_left), str(max_ee), str(truncq),
-               temp_dir_name]
+        cmd = ['run_dada_faster.R',
```
Comment (gregcaporaso) [inline q2_dada2/_denoise.py]: Should we change the name of the R script to `run_data_single.R` (and `run_data_paired.R`)?

File: q2_dada2/_denoise.py (hunk 3 above)
Comment (gregcaporaso) [inline q2_dada2/_denoise.py]: *arguments* to *truncation and trim parameter settings*

File: q2_dada2/_denoise.py [hunk 5]
```diff
@@ -6,28 +6,93 @@
...
+                       for id_ in table.ids(axis='observation')}
+            table.update_ids(fid_map, axis='observation', inplace=True)
+
+            rep_sequences = DNAIterator((skbio.DNA(k, metadata={'id': v})
+                                         for k, v in fid_map.items()))
+        else:
+            rep_sequences = DNAIterator(
+                (skbio.DNA(id_, metadata={'id': id_})
+                 for id_ in table.ids(axis='observation')))
+    return table, rep_sequences
```
Comment (gregcaporaso) [inline q2_dada2/_denoise.py]: Can lines 44 - 64 be moved to a single function that can be called from `denoise_single` and `denoise_paired`?

File: q2_dada2/plugin_setup.py [hunk 6]
```diff
@@ -32,25 +34,59 @@
...
+                'trunc_q': qiime2.plugin.Int,
+                'n_threads': qiime2.plugin.Int,
+                'n_reads_learn': qiime2.plugin.Int,
                 'hashed_feature_ids': qiime2.plugin.Bool},
     outputs=[('table', FeatureTable[Frequency]),
              ('representative_sequences', FeatureData[Sequence])],
-    name='Denoise and dereplicate',
-    description=('This method denoises sequences, dereplicates them, and '
-                 'filters chimeras.')
+    name='Denoise and dereplicate single-end',
```
Comment (gregcaporaso) [inline q2_dada2/plugin_setup.py]: *single-end sequences.*

**Pull request merged**